File: leetCode/fukun/201-300/lc_238.js
var productExceptSelf = function (nums) {
    let output = [1];
    for (let i = 1; i < nums.length; i++) {
        output[i] = nums[i - 1] * (output[i - 1]);
    }
    for (let i = nums.length - 2; i >= 0; i--) {
        output[i] *= nums[i + 1];
        nums[i] *= nums[i + 1];
    }
    return output;
};

console.log(productExceptSelf([1, 2, 3, 4]));

File: leetCode/lwc/0201-0300/lc_238.java
package leetcode;

/**
 * @author lwc
 * @date 2019/11/25 18:09
 */
public class lc_238 {
    public int[] productExceptSelf(int[] nums) {
        int[] v = new int[nums.length];
        int pos = -1;
        int total = 1;
        for(int i = 0; i < nums.length; i++){
            if(nums[i] == 0){
                if(pos > -1) return v;
                else pos = i;
            }
            else total *= nums[i];
        }
        if(pos != -1) v[pos] = total;
        else{
            for(int i = 0; i < nums.length; i++){
                v[i] = total / nums[i];
            }
        }
        return v;
    }
}
